compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/CollectionStubMethodLowering.kt

```diff
@@ -0,0 +1,274 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.backend.jvm.lower
+
+import org.jetbrains.kotlin.backend.common.ClassLoweringPass
+import org.jetbrains.kotlin.backend.common.descriptors.WrappedValueParameterDescriptor
+import org.jetbrains.kotlin.backend.common.lower.createIrBuilder
+import org.jetbrains.kotlin.backend.common.lower.irThrow
+import org.jetbrains.kotlin.backend.common.phaser.makeIrFilePhase
+import org.jetbrains.kotlin.backend.jvm.JvmBackendContext
+import org.jetbrains.kotlin.backend.jvm.util.eraseTypeParameters
+import org.jetbrains.kotlin.descriptors.Modality
+import org.jetbrains.kotlin.ir.UNDEFINED_OFFSET
+import org.jetbrains.kotlin.ir.builders.declarations.buildFun
+import org.jetbrains.kotlin.ir.builders.irBlockBody
+import org.jetbrains.kotlin.ir.builders.irCall
+import org.jetbrains.kotlin.ir.builders.irString
+import org.jetbrains.kotlin.ir.declarations.*
+import org.jetbrains.kotlin.ir.declarations.impl.IrValueParameterImpl
+import org.jetbrains.kotlin.ir.symbols.IrClassSymbol
+import org.jetbrains.kotlin.ir.symbols.IrTypeParameterSymbol
+import org.jetbrains.kotlin.ir.symbols.impl.IrValueParameterSymbolImpl
+import org.jetbrains.kotlin.ir.types.*
+import org.jetbrains.kotlin.ir.types.impl.IrSimpleTypeImpl
+import org.jetbrains.kotlin.ir.types.impl.makeTypeProjection
+import org.jetbrains.kotlin.ir.util.*
+import org.jetbrains.kotlin.name.FqName
+
+internal val collectionStubMethodLowering = makeIrFilePhase(
+    ::CollectionStubMethodLowering,
+    name = "CollectionStubMethod",
+    description = "Generate Collection stub methods"
+)
+
+private class CollectionStubMethodLowering(val context: JvmBackendContext) : ClassLoweringPass {
+
+    private val state = context.state
+
+    private val typeMapper = state.typeMapper
+
+    override fun lower(irClass: IrClass) {
+        if (irClass.isInterface) {
+            return
+        }
+
+        val methodStubsToGenerate = findRelevantStubMethods(irClass)
+
+        // We don't need to generate stub for existing methods, but for FAKE_OVERRIDE methods with ABSTRACT modality,
+        // it means an abstract function in superclass that is not implemented yet,
+        // stub generation is still needed to avoid invocation error.
+        val existingMethodSignatures = irClass.functions.filter {
+            it.modality != Modality.ABSTRACT || it.origin != IrDeclarationOrigin.FAKE_OVERRIDE
+        }.associate { it.toSignature() to it }.toMap()
```

Comment:
`.associate { it.toSignature() to it }` -> `.associateBy { it.toSignature() }`

Also consider using `filterNot` instead of `filter` above